manifests/07-operator.yaml

```diff
@@ -80,3 +83,12 @@ spec:
             items:
             - key: ca-bundle.crt
               path: tls-ca-bundle.pem
+        # This service account token can be used to provide identity outside the cluster.
+        # For example, this token can be used with AssumeRoleWithWebIdentity to authenticate
+        # with AWS using an IAM OIDC provider and STS.
+        - name: bound-sa-token
+          projected:
+            sources:
+            - serviceAccountToken:
+                path: token
+                audience: openshift
```

Comment:
This is what makes the entire web of trust work between AWS and the Pod running inside the cluster. You configure an AWS IAM Identity Provider that trusts identities that the OpenShift cluster can validate/sign. So now the Pod can make an iam:STS API call and say "here is my token signed by an OpenID provider that you trust". Then AWS validates that the token is indeed signed by something that it trusts, and (assuming other things check out properly) AWS will hand back temporary credentials.

Reply:
Ah, I understand now. Thanks!

Reply:
@joelddiaz Isn't this the token for the `cluster-image-registry-operator` service account? If this is the case do we still need to explicitly mount it? As we are not explicitly opting out of `automountServiceAccountToken` the token is already mounted under `/var/run/secrets/kubernetes.io/serviceaccount`. Please clarify.

Reply:
The contents of the two ServiceAccounts are slightly different.

Possibly most importantly the keyId field in the kubernetes.io/serviceaccount location doesn't match what is uploaded to the OIDC S3 bucket:

Standard ServiceAccount token:
`sh-4.4$ cat /var/run/secrets/kubernetes.io/serviceaccount/token | awk -F. '{ print $1 }' | base64 -d
{"alg":"RS256","kid":"RjUCNiAxtfGZcTjdVWLvNyksvfhTuQ7dmqAxBrbZIYA"}base64: invalid input
`
Projected ServiceAccount token:
`sh-4.4$ cat /var/run/secrets/openshift/serviceaccount/token | awk -F. '{ print $1 }' | base64 -d
{"alg":"RS256","kid":"lQZjBZuL5Xzq0B6Ih3VQVWFi6zVPE8u7fq_yubth2GQ"}base64: invalid input
`

And looking at the OIDC keys.json, you see the missmatch:

```
[jdiaz@minigoomba ~]$ curl https://jdiaz-stsfull-installer.s3.amazonaws.com/keys.json | jq -r '.keys[].kid'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   917  100   917    0     0   2345      0 --:--:-- --:--:-- --:--:--  2351
lQZjBZuL5Xzq0B6Ih3VQVWFi6zVPE8u7fq_yubth2GQ
```

Looking at the second field of the token there are more differences:

Standard ServiceAccount token:
`sh-4.4$ cat /var/run/secrets/kubernetes.io/serviceaccount/token | awk -F. '{ print $2 }' | base64 -d
{"iss":"kubernetes/serviceaccount","kubernetes.io/serviceaccount/namespace":"openshift-image-registry","kubernetes.io/serviceaccount/secret.name":"cluster-image-registry-operator-token-jbtdj","kubernetes.io/serviceaccount/service-account.name":"cluster-image-registry-operator","kubernetes.io/serviceaccount/service-account.uid":"30278254-d1bc-4631-ad41-777cbef7bb39","sub":"system:serviceaccount:openshift-image-registry:cluster-image-registry-operator"}base64: invalid input`

Projected ServiceAccount:
`sh-4.4$ cat /var/run/secrets/openshift/serviceaccount/token | awk -F. '{ print $2 }' | base64 -d
{"aud":["openshift"],"exp":1611938123,"iat":1611934523,"iss":"https://s3.us-east-1.amazonaws.com/jdiaz-stsfull-installer","kubernetes.io":{"namespace":"openshift-image-registry","pod":{"name":"cluster-image-registry-operator-8cb64687b-bpdp9","uid":"5050f308-9c1d-4c3c-89a4-8c454d0e08f6"},"serviceaccount":{"name":"cluster-image-registry-operator","uid":"30278254-d1bc-4631-ad41-777cbef7bb39"}},"nbf":1611934523,"sub":"system:serviceaccount:openshift-image-registry:cluster-image-registry-operator"}`

Here the issuer field (iss) differs, there is a missing audience (aud), no expiration (exp) or issued-at (iat).

But I do agree that they are fundamentally the same as you can see the serviceaccount.uid match.

Reply:
Thanks for clarifying this aspect. Two distinct tokens indeed :-)

Reply:
Is this automatically set, or does the user need to do anything manually?

Reply:
The Deployment for the image-registry-operator is permanently changed to start projecting in the ServiceAccount to this /var/run/secrets/openshift/serviceaccount/token location. OpenShift handles signing the token and updating/rotating it as needed automatically. Whether or not we use the ServiceAccount token, it will be there going forward (even for non-AWS clusters).